### DISCUSSION TITLE
Mouse event listener issue

### DISCUSSION THREAD

**Discussion by Abhay0294** [eclipse-glsp/glsp#705] [Q&A] [answered]
HI,

I have created an extension which will be returning the mouse position by extending a MouseTool class.

**Below is the code for the extension:**
`@injectable()
export class HideNodeAction extends MouseTool {

    @inject(MousePositionTracker) protected readonly mousePosition: MousePositionTracker;

    decorate(vnode: VNode, element: SModelElement): VNode {
        if (element instanceof SModelRoot) {
            on(vnode, 'dblclick', this.doubleClick.bind(this, element));
        }
        return vnode;
    }

    doubleClick(element: SModelRoot, event: MouseEvent): void {
        this.getMousePosition(element, event);
    }

    getMousePosition(element: SModelRoot, event: MouseEvent): void {
        const x = new MousePositionTracker();
        x.mouseMove(element, event);
        console.log('Your mouse position is:', x);
    }

}`

**and below is di.config for the extension:**
`const hideNodeActionModule = new ContainerModule((bind, _unbind, _isBound, rebind) => {

    bind(HideNodeAction).toSelf().inSingletonScope();
    _unbind(MouseTool);
    bind(MouseTool).to(HideNodeAction);
});`

It is working fine, I'm able to get the position of the mouse when I dbClick anywhere on the editor.

But the issue is that I'm now unable to perform any action on the editor like selecting the node, dragging nodes here and there, etch.

Please help me by providing the feedback on what I'm doing wrong.

**Comment by martin-fleck-at** [comment, marked as answer]
Hi @Abhay0294 

Please note that in GLSP we already use a custom mouse tool, the [`RankingMouseTool`](https://github.com/eclipse-glsp/glsp-client/blob/master/packages/client/src/features/mouse-tool/mouse-tool.ts) if you use the [`glspMouseToolModule`](https://github.com/eclipse-glsp/glsp-client/blob/master/packages/client/src/features/mouse-tool/di.config.ts). In that case it would be better for you to rebind that as well instead of only the MouseTool.

I think in your case registering your action as a mouse listener and injecting the `MousePositionTracker` (for the `lastPositionOnDiagram`) should be enough. The original mouse tool singleton will notify all registered mouse listeners and already has a proper binding for double click. So you can simply do `bind(TYPES.MouseListener).to(HideNodeAction)` and implement that interface instead. However, if you really do rebind the mouse tool, you probably need to call `super.decorate(vnode, element)` in your method to get all the other default bindings as well (untested).

**Reply by Abhay0294** [reply]
Should  I call super.decorate(vnode, element) in my decorate method ?

**Reply by martin-fleck-at** [reply]
If you really need to override the singleton mouse tool, then yes, I'd suggest to call the super decorate method in your decorate method as it creates the bindings for all the mouse events and their respective methods. Any registered mouse listener is then notified about any of those events.

**Reply by Abhay0294** [reply]
@martin-fleck-at  is it correct to rebind RankingMouseTool as you mentioned?
`rebind(MouseTool).to(RankingMouseTool);`

**Reply by martin-fleck-at** [reply]
If you use the glspMouseToolModule in your application then the RankingMouseTool is already correctly bound as the "default" mouse tool, see https://github.com/eclipse-glsp/glsp-client/blob/master/packages/client/src/features/mouse-tool/di.config.ts. I think what you need is to create a mouse listener that reacts to the double click and retrieves the current mouse position from the MousePositionTracker.

**Reply by Abhay0294** [reply]
Thanks a lot @martin-fleck-at
